libpgcommon/lwgeom_transform.c

```diff
@@ -451,102 +521,164 @@ static char* GetProj4String(int srid)
 		}
 		else
 		{
 			elog(ERROR, "Invalid reserved SRID (%d)", srid);
-			return NULL;
+			return strs;
 		}
 
-		POSTGIS_DEBUGF(3, "returning on SRID=%d: %s", srid, proj_str);
-		return proj_str;
+		POSTGIS_DEBUGF(3, "returning on SRID=%d: %s", srid, strs.proj4text);
+		return strs;
 	}
 }
 
-void AddToPROJ4Cache(Proj4Cache cache, int srid, int other_srid) {
-	AddToPROJ4SRSCache((PROJ4PortalCache *)cache, srid, other_srid) ;
+static int
+pjstrs_has_entry(const PjStrs *strs)
+{
+	if ((strs->proj4text && strlen(strs->proj4text)) ||
+		(strs->epsgtext && strlen(strs->epsgtext)) ||
+		(strs->srtext && strlen(strs->srtext)))
+		return 1;
+	else
+		return 0;
 }
 
+static void
+pjstrs_pfree(PjStrs *strs)
+{
+	if (strs->proj4text)
+		pfree(strs->proj4text);
+	if (strs->epsgtext)
+		pfree(strs->epsgtext);
+	if (strs->srtext)
+		pfree(strs->srtext);
+}
+
+static char*
+pgstrs_get_entry(const PjStrs *strs, int n)
+{
+	switch (n)
+	{
+		case 0:
+			return strs->epsgtext;
+		case 1:
+			return strs->srtext;
+		case 2:
+			return strs->proj4text;
+		default:
+			return NULL;
+	}
+}
 
 /**
- * Add an entry to the local PROJ4 SRS cache. If we need to wrap around then
+ * Add an entry to the local PROJ SRS cache. If we need to wrap around then
  * we must make sure the entry we choose to delete does not contain other_srid
  * which is the definition for the other half of the transformation.
  */
 static void
-AddToPROJ4SRSCache(PROJ4PortalCache *PROJ4Cache, int srid, int other_srid)
+AddToPROJSRSCache(PROJPortalCache *PROJCache, int srid_from, int srid_to)
 {
 	MemoryContext PJMemoryContext;
-	projPJ projection = NULL;
-	char *proj_str = NULL;
+
+	PjStrs from_strs;
+	PjStrs to_strs;
 
 	/*
 	** Turn the SRID number into a proj4 string, by reading from spatial_ref_sys
 	** or instantiating a magical value from a negative srid.
 	*/
-	proj_str = GetProj4String(srid);
-	if ( ! proj_str )
+	from_strs = GetProjStrings(srid_from);
+	to_strs = GetProjStrings(srid_to);
+	if (!pjstrs_has_entry(&from_strs))
+		elog(ERROR, "AddToPROJSRSCache: GetProjStrings returned NULL for SRID (%d)", srid_from);
+	if (!pjstrs_has_entry(&to_strs))
+		elog(ERROR, "AddToPROJSRSCache: GetProjStrings returned NULL for SRID (%d)", srid_to);
+
+#if POSTGIS_PROJ_VERSION < 60
+	PJ* projection = malloc(sizeof(PJ));
+	projection->pj_from = lwproj_from_string(from_strs.proj4text);
+	projection->pj_to = lwproj_from_string(to_strs.proj4text);
+
+	if (!projection->pj_from)
+		elog(ERROR,
+		    "AddToPROJSRSCache: could not form projection from 'srid=%d' to 'srid=%d'",
+		    srid_from, srid_to);
+
+	if (!projection->pj_to)
+		elog(ERROR,
+		    "AddToPROJSRSCache: could not form projection from 'srid=%d' to 'srid=%d'",
+		    srid_from, srid_to);
+#else
+	PJ* projection = NULL;
+	/* Try combinations of ESPG/SRTEXT/PROJ4TEXT until we find */
+	/* one that gives us a usable transform. Note that we prefer */
+	/* EPSG numbers over SRTEXT and SRTEXT over PROJ4TEXT */
+	for (uint32_t i = 0; i < 9; i++)
```

Comment:
It'd nice if the comment explained that 9 comes from 3^2. I know it says it more or less, but at first I thought it was `PROJ_CACHE_ITEMS` + 1-

BTW, since this is compiled with Postgres flags, expect compiler complaints about declaring variables inside the `for` for Postgres releases before 11, for example this will break with PG10+PROJ6+Werror. `been there, done that`